### PR TITLE
fix(Anchor): fix highlighting failure when dynamically updating items

### LINKED ISSUE #40730: anchor 使用動態 items 會產生問題
### Reproduction link

[https://youtu.be/Hssq9pySchU](https://youtu.be/Hssq9pySchU)

### Steps to reproduce

1. 點擊Anchor锚点頁 隐藏调试专用演示
2. 滑動內容

### What is expected?

正常對應到anchor item

### What is actually happening?

anchor顯示會卡住

| Environment | Info |
| --- | --- |
| antd | 5.2.1 |
| React | 18.2.0 |
| System | mac |
| Browser | Chrome 108.0.5359.124 |

---
看起來是容器內容高度改變時沒有重新計算錨點高度
我是使用react-hook-form內的field array動態產生anchor items和container內容

[https://legacy.react-hook-form.com/api/usefieldarray](https://legacy.react-hook-form.com/api/usefieldarray)

list 1
list 2
list 3
  child item 1  
  child item 2 <-- 動態新增 item3
list 4
list 5
list 6

測試結果為如果新增內容時 該父項(list 3)為高亮時，新增後正常
若此時高亮為其他項目(list 2)時 新增內容後anchor會卡住。

### PR DESCRIPTION
**Pull request by zqran**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [ ] New feature
- [x] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link

- close #40730

### 💡 Background and solution

[复现地址](https://codesandbox.io/s/ji-ben-antd-5-2-1-forked-u5kng8?file=/demo.tsx)

复现步骤：

1. 进入页面后，点击右上角按钮，新增一个 anchor 节点
2. 滚动页面到底部，出现第一个 Bug：最后一个节点未高亮
3. 滚动回到页面顶部，再次点击右上角按钮，滚动页面，此时出现第二个 Bug：anchor 节点不再高亮

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |      fix highlighting failure when dynamically updating items     |
| 🇨🇳 Chinese |      修复动态添加 items 后高亮失效问题     |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

### CODE REVIEW

**Comment by MadCcc** [comment]
来个测试用例？

File: components/anchor/Anchor.tsx [hunk 1]
```diff
@@ -154,24 +154,20 @@ const AnchorContent: React.FC<InternalAnchorProps> = (props) => {
   const getCurrentContainer = getContainer ?? getTargetContainer ?? getDefaultContainer;
 
   const dependencyListItem: React.DependencyList[number] = JSON.stringify(links);
+  const linksRef = React.useRef(links);
+  linksRef.current = links;
```
Comment (MadCcc) [inline components/anchor/Anchor.tsx:158]: 这样每次渲染 ref 都会和 state 同步？那是不是直接用 links 就好了

**Comment by MadCcc** [comment]
一个经典 useEvent 场景

**Comment by li-jia-nan** [comment]
看样子是我之前重构 FC 改出问题了……

**Comment by zqran** [comment]
> 看样子是我之前重构 FC 改出问题了……

React 的依赖项，还是挺坑的~

**Comment by zqran** [comment]
> 看样子是我之前重构 FC 改出问题了……

React 的依赖项，还是挺坑的~

**Comment by afc163** [comment]
来个测试用例？

**Comment by zqran** [comment]
> 来个测试用例？

@MadCcc 求助攻~

**Comment by zqran** [comment]
> 来个测试用例？

@MadCcc 求助攻~

**Comment by frankA10019** [comment]
any update?

**Review by li-jia-nan — APPROVED** [review]

**Comment by li-jia-nan** [comment]
本地试了一下没有问题，先 merge 了，测试用例我单独开个 PR 补一下吧

**Comment by li-jia-nan** [comment]
> 本地试了一下没有问题，先 merge 了，测试用例我单独开个 PR 补一下吧

- #41162

**Comment by pc110** [comment]
> 看样子是我之前重构 FC 改出问题了……

Bug 最近很多哦，有很多很表面的bug哟， 注意下代码质量哟

**Pull request merged**